In [1]:
# import the libraries needed
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

## Preparing the train and test dataset
---
- We have obtained the dataset from https://www.openpowerlifting.org/
- From there, we randomised the data rows in the original CSV file and took out a total of 3,000 data points to be used for our train and test dataset

In [2]:
train_test_combined = pd.read_csv('../datasets/train_test_combined.csv')
train_test_combined.head()

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
0       Alana Carrasco   F   SBD         Raw  29.5    24-34          24-39   
1       Even Dysjaland   M   SBD         Raw  19.5    20-23          19-23   
2  Alessandro Favorito   M   SBD  Single-ply  39.0    35-39          24-39   
3      Sydney Martinez   F   SBD  Single-ply   NaN      NaN            NaN   
4          Navy Villar   F   SBD  Single-ply   NaN      NaN            NaN   

        Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
0           FR-O         68.60            72  ...     Yes      USA     AZ   
1  Juniors 19-23         94.65           105  ...     Yes      NaN    NaN   
2           Open         89.40            90  ...     Yes    Italy    NaN   
3          Girls         55.16            56  ...     Yes      USA    NaN   
4          Girls         55.61            56  ...     Yes      NaN    NaN   

   Federation  ParentFederation       Date  MeetCountry  MeetState  \
0       USAPL               IPF  27/8/2017          USA         AZ   
1         NSF               IPF  13/6/2015       Norway        NaN   
2        FIPL               IPF  6/11/2004        Italy        NaN   
3      THSWPA               NaN  25/1/2014          USA         TX   
4      THSWPA               NaN  18/1/2018          USA         TX   

            MeetTown                      MeetName  
0                NaN             Sun Devil Classic  
1            Ganddal                   Klubbstevne  
2  Marina Di Carrara        Coppa Italia A Squadre  
3        JUNIOR HIGH  CARRIZO SPRINGS INVITATIONAL  
4             Sadler            S&S Lady Rams Meet  

[5 rows x 41 columns]

In [8]:
# Split combined dataset into train and test
from sklearn.model_selection import train_test_split

train,test = train_test_split(train_test_combined, test_size = 0.25, random_state = 0)
train.to_csv('../datasets/train.csv', index = False)
test.to_csv('../datasets/test.csv', index = False)

print(train.shape)
print(test.shape)

(3000, 41)
(1000, 41)


In [9]:
# Import train dataset into a dataframe
train_df = pd.read_csv('../datasets/train.csv')
train_df.head()

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
0    Hanne Bingle   F   SBD       Wraps  61.0    60-64          60-69   
1    Mohamed Reda   M   SBD         Raw   NaN      NaN            NaN   
2  Pauline Reeves   F   SBD  Single-ply   NaN    45-49            NaN   
3   Edie Montalvo   F   SBD  Single-ply   NaN      NaN            NaN   
4    Faith Bailey   F   SBD  Single-ply   NaN      NaN            NaN   

  Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
0       M5         73.70            75  ...     NaN       UK    NaN   
1     Open         99.58           105  ...     Yes    Egypt    NaN   
2     Open        129.30           90+  ...     Yes  England    NaN   
3    Girls         71.12            75  ...     Yes      NaN    NaN   
4    Girls         78.83          82.3  ...     Yes      NaN    NaN   

   Federation  ParentFederation        Date  MeetCountry  MeetState  \
0         GPC               GPC    1/9/2020     Slovakia        NaN   
1   AfricanPF               IPF  20/10/2018      Morocco        NaN   
2       BAWLA               IPF   26/4/2003           UK        NaN   
3      THSWPA               NaN   10/2/2018          USA         TX   
4      THSWPA               NaN    7/3/2020          USA         TX   

      MeetTown                                      MeetName  
0       Trnava                        European Championships  
1       Meknes     African & Arab Powerlifting Championships  
2  Livingstone                British Masters' Championships  
3      Weslaco  Weslaco Girls Invitational Powerlifting Meet  
4     Bay City                      Region 4 Division 3 Meet  

[5 rows x 41 columns]

## Data Preprocessing
---

### Removing disqualified lifters

First, find out if any lifters were disqualified (DQ) or did not turn up for the competition (NS). Proceed to remove them from our dataset.

In [13]:
train_df['Place'].unique()

array(['1', '3', 'DQ', '7', '9', '2', '4', '6', 'NS', '5', '14', '11',
       '10', '8', '13', '12', '18', '15', '22', '24', '32', '26', '17',
       '19', '16', 'G', '28', 'DD', '23', '51', '21', '84', '20', '29'],
      dtype=object)

In [14]:
df = train_df.drop(train_df[
    (train_df.Place == 'DQ') |
    (train_df.Place == 'NS') |
    (train_df.Place == 'DD')
].index)

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2814 non-null   object 
 1   Sex               2814 non-null   object 
 2   Event             2814 non-null   object 
 3   Equipment         2814 non-null   object 
 4   Age               1613 non-null   float64
 5   AgeClass          1884 non-null   object 
 6   BirthYearClass    1721 non-null   object 
 7   Division          2812 non-null   object 
 8   BodyweightKg      2775 non-null   float64
 9   WeightClassKg     2788 non-null   object 
 10  Squat1Kg          1284 non-null   float64
 11  Squat2Kg          1270 non-null   float64
 12  Squat3Kg          1230 non-null   float64
 13  Squat4Kg          11 non-null     float64
 14  Best3SquatKg      2787 non-null   float64
 15  Bench1Kg          1285 non-null   float64
 16  Bench2Kg          1277 non-null   float64


After dropping disqualified and no-show lifters from the dataset, we are left with only 2814 values from the original 3000 values. 

## Filling Missing Values (Age)
First, use the AgeClass column to fill in missing values for Age column. We will assume that the lifter's age is the maximum age within the AgeClass.

In [7]:
df[df['Age'].isna() & df['AgeClass'].notna()]

Name Sex Event   Equipment  Age AgeClass BirthYearClass  \
21      Melissa Salinas   F   SBD  Single-ply  NaN    16-17          14-18   
26          Amy Hoffman   F   SBD  Single-ply  NaN    45-49          40-49   
33          Nick Cooper   M   SBD  Single-ply  NaN    13-15          14-18   
40        Atle Andersen   M   SBD  Single-ply  NaN    24-34          24-39   
56         Diane Farrar   F   SBD  Single-ply  NaN    45-49          40-49   
...                 ...  ..   ...         ...  ...      ...            ...   
3928         J. Shelton   M   SBD  Single-ply  NaN    16-17          14-18   
3943   Sanjuana Delgado   F   SBD  Single-ply  NaN    18-19            NaN   
3945         Ramirez Jr   M   SBD  Single-ply  NaN    40-44          40-49   
3965  Kayla Anderson #3   F   SBD  Single-ply  NaN    18-19            NaN   
3975    Taylor Allen #1   F   SBD  Single-ply  NaN    16-17            NaN   

           Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
21            Girls         73.39            75  ...     Yes      NaN    NaN   
26    Masters 45-49         75.00            75  ...     NaN      NaN    NaN   
33             M-T1         75.00            75  ...     Yes      USA    NaN   
40             Open         74.60            75  ...     Yes      NaN    NaN   
56            F-M1b         60.00            60  ...     Yes      NaN    NaN   
...             ...           ...           ...  ...     ...      ...    ...   
3928     Teen 14-16         67.50          67.5  ...     Yes   Canada    NaN   
3943          Girls         55.97            56  ...     Yes      NaN    NaN   
3945  Masters 40-44         90.00            90  ...     NaN      NaN    NaN   
3965          Girls         93.71           100  ...     Yes      NaN    NaN   
3975          Girls         60.33          67.3  ...     Yes      NaN    NaN   

      Federation  ParentFederation       Date  MeetCountry  MeetState  \
21        THSWPA               NaN  17/1/2015          USA         TX   
26          USPF               NaN  5/12/2015          USA         WV   
33         USAPL               IPF  23/3/2002          USA         FL   
40           NSF               IPF  24/2/1991       Norway        NaN   
56         USAPL               IPF  15/5/2004          USA         RI   
...          ...               ...        ...          ...        ...   
3928       ADFPA               NaN  23/3/1990          USA         MS   
3943      THSWPA               NaN   5/3/2016          USA         TX   
3945        USPF               IPF  14/5/1993          USA         TX   
3965      THSWPA               NaN  24/1/2019          USA         TX   
3975      THSWPA               NaN  18/2/2017          USA         TX   

             MeetTown                              MeetName  
21     Skidmore Texas      2nd Annual Skidmore Invitational  
26        Parkersburg                Region 2 Championships  
33           Lakeland                 Frank Kostyo Memorial  
40    Lykkeberghallen                           Seriestevne  
56            Warwick                    New England States  
...               ...                                   ...  
3928          Jackson                   Mississippi Teenage  
3943          Academy                   Region 2 Division 3  
3945           Irving                     Masters Nationals  
3965        Henderson                     Lion Invitational  
3975        WHITE OAK  EAST TEXAS POWERLIFTING CHAMPIONSHIP  

[339 rows x 41 columns]

In [179]:
def newString(g):
    ## Function that returns the last two characters of a str as an int. 
    if type(g) == str:
        return int(g[-2:])
    return g

In [180]:
## First, fill the values using the AgeClass Strings
df['Age'].fillna(df.AgeClass, inplace=True)

In [181]:
## Use the newString function to replace the Strings with the maximum age 
df['Age'] = df['Age'].apply(newString)

In [183]:
#compare indexes against the DataFrame printed earlier to ensure the replacement was done correctly
df.loc[29]

Name                       Ashley Gilbert
Sex                                     F
Event                                 SBD
Equipment                      Single-ply
Age                                  19.0
AgeClass                            18-19
BirthYearClass                        NaN
Division                            Girls
BodyweightKg                        79.56
WeightClassKg                        82.3
Squat1Kg                              NaN
Squat2Kg                              NaN
Squat3Kg                              NaN
Squat4Kg                              NaN
Best3SquatKg                       102.06
Bench1Kg                              NaN
Bench2Kg                              NaN
Bench3Kg                              NaN
Bench4Kg                              NaN
Best3BenchKg                        72.57
Deadlift1Kg                           NaN
Deadlift2Kg                           NaN
Deadlift3Kg                           NaN
Deadlift4Kg                       

In [121]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2828 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2828 non-null   object 
 1   Sex               2828 non-null   object 
 2   Event             2828 non-null   object 
 3   Equipment         2828 non-null   object 
 4   Age               2228 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2118 non-null   object 
 7   Division          2828 non-null   object 
 8   BodyweightKg      2822 non-null   float64
 9   WeightClassKg     2813 non-null   object 
 10  Squat1Kg          1085 non-null   float64
 11  Squat2Kg          1077 non-null   float64
 12  Squat3Kg          1054 non-null   float64
 13  Squat4Kg          5 non-null      float64
 14  Best3SquatKg      1871 non-null   float64
 15  Bench1Kg          1712 non-null   float64
 16  Bench2Kg          1703 non-null   float64


Now the age column only contains numeric float64 datatypes, which shows that we have successfully replaced all the strings after the fillna() operation.

However, there are still some missing values, where both Age and AgeClass data was missing.

We can fill the remaining missing values for Age with the median age of their respective divisions.

In [184]:
df[df['Age'].isna()]

Name Sex Event   Equipment  Age AgeClass  \
2                 Paige Shuemake   F   SBD  Single-ply  NaN      NaN   
3                     Pål Nilsen   M   SBD  Single-ply  NaN      NaN   
8                      Tori Rash   F   SBD  Single-ply  NaN      NaN   
9     Stina Christiane Valmestad   F     B         Raw  NaN      NaN   
10                Kyle Monhollen   M   SBD         Raw  NaN      NaN   
...                          ...  ..   ...         ...  ...      ...   
2980            Alejandra Fierro   F   SBD  Single-ply  NaN      NaN   
2986          Tor Erik Rødsdalen   M   SBD  Single-ply  NaN      NaN   
2995             Roberto Bettati   M   SBD  Single-ply  NaN      NaN   
2996            Lachlan Hodgetts   M     B         Raw  NaN      NaN   
2999                Paul Gallant   M     B  Single-ply  NaN      NaN   

     BirthYearClass   Division  BodyweightKg WeightClassKg  ...  Tested  \
2               NaN      Girls         54.25            56  ...     Yes   
3               NaN       Open         89.20            90  ...     Yes   
8               NaN      Girls         59.96          60.1  ...     Yes   
9               NaN       Open         73.23            84  ...     Yes   
10              NaN       MR-O        104.20           105  ...     Yes   
...             ...        ...           ...           ...  ...     ...   
2980            NaN      Girls         51.94          51.9  ...     Yes   
2986            NaN       Open         89.40            90  ...     Yes   
2995          40-49  Masters 1         88.40            90  ...     Yes   
2996            NaN       Open         72.70            74  ...     Yes   
2999            NaN       Open         90.00            90  ...     Yes   

      Country  State  Federation  ParentFederation        Date  MeetCountry  \
2         NaN    NaN      THSWPA               NaN   31/1/2015          USA   
3      Norway    NaN         NSF               IPF    6/6/1998       Norway   
8         NaN    NaN      THSWPA               NaN   24/1/2019          USA   
9         NaN    NaN         NSF               IPF  15/12/2018       Norway   
10        NaN    NaN       USAPL               IPF   22/8/2015          USA   
...       ...    ...         ...               ...         ...          ...   
2980      NaN    NaN      THSWPA               NaN    7/2/2019          USA   
2986      NaN    NaN         NSF               IPF   16/2/2002       Norway   
2995    Italy    NaN        FIPL               IPF   29/4/2006        Italy   
2996   Canada     BC         CPU               IPF   23/6/2018       Canada   
2999   Canada     NB         CPU               IPF   8/11/1997       Canada   

      MeetState       MeetTown                           MeetName  
2            TX       Malakoff                           Malakoff  
3           NaN           Oslo                   KM Oslo-Akershus  
8            TX      Needville                  Needville 1-24-19  
9           NaN          Sande                        Klubbstevne  
10           OH            NaN            Future Summer Shakedown  
...         ...            ...                                ...  
2980         TX      Plainview               Plainview Triangular  
2986        NaN     Brumunddal                    B-og Veteran NM  
2995        NaN        Nettuno       Campionato Italiano Assoluto  
2996         BC         Surrey       BCPA Provincial Championship  
2999         PE  Charlottetown  Eastern Canadian BP Championships  

[600 rows x 41 columns]

In [217]:
t = df.groupby('Division')['Age'].median()
t ## Series showing median age of each division

Division
13-14               13.5
17-19               18.0
40-49               45.0
45-49               48.0
55-59               57.0
                    ... 
Youth 10-15         13.5
Youth 11 & Under     6.0
Youth 11-12         11.0
sen                  NaN
Юниорки              NaN
Name: Age, Length: 369, dtype: float64

In [216]:
df['Age'] = df['Age'].fillna(df.groupby('Division')['Age'].transform('median'))

##stackoverflow magic https://stackoverflow.com/questions/19966018/pandas-filling-missing-values-by-mean-in-each-group


In [221]:
## cell for checking
print(df.loc[8]['Age']) ## value in the Age column after the fillna()
print(t[df.loc[8]['Division']]) ## median value of the division

19.0
19.0


In [222]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2828 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2828 non-null   object 
 1   Sex               2828 non-null   object 
 2   Event             2828 non-null   object 
 3   Equipment         2828 non-null   object 
 4   Age               2793 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2118 non-null   object 
 7   Division          2828 non-null   object 
 8   BodyweightKg      2822 non-null   float64
 9   WeightClassKg     2813 non-null   object 
 10  Squat1Kg          1085 non-null   float64
 11  Squat2Kg          1077 non-null   float64
 12  Squat3Kg          1054 non-null   float64
 13  Squat4Kg          5 non-null      float64
 14  Best3SquatKg      1871 non-null   float64
 15  Bench1Kg          1712 non-null   float64
 16  Bench2Kg          1703 non-null   float64


Drop the remaining values NA values for age

In [227]:
df.dropna(subset=['Age'], inplace=True)

In [229]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2793 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2793 non-null   object 
 1   Sex               2793 non-null   object 
 2   Event             2793 non-null   object 
 3   Equipment         2793 non-null   object 
 4   Age               2793 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2114 non-null   object 
 7   Division          2793 non-null   object 
 8   BodyweightKg      2790 non-null   float64
 9   WeightClassKg     2780 non-null   object 
 10  Squat1Kg          1074 non-null   float64
 11  Squat2Kg          1066 non-null   float64
 12  Squat3Kg          1044 non-null   float64
 13  Squat4Kg          4 non-null      float64
 14  Best3SquatKg      1850 non-null   float64
 15  Bench1Kg          1696 non-null   float64
 16  Bench2Kg          1687 non-null   float64


# Filling Missing Values (Bodyweight)

In [231]:
df[df.BodyweightKg.isna()]

Name Sex Event   Equipment   Age AgeClass BirthYearClass  \
1086   Richard Nowazek   M   SBD  Single-ply  51.5    50-54          50-59   
1597  Madison Neesmith   F     B         Raw  20.0    20-23          19-23   
1605         Jay Bakke   M   SBD   Multi-ply  53.0    50-54          50-59   

          Division  BodyweightKg WeightClassKg  ...  Tested  Country  State  \
1086          Open           NaN          140+  ...     Yes   Canada     BC   
1597            JR           NaN         82.5+  ...     Yes      USA     NC   
1605  M_MEM_3_AAPF           NaN           NaN  ...     Yes      USA    NaN   

      Federation  ParentFederation       Date  MeetCountry  MeetState  \
1086         CPU               IPF   6/4/2002       Canada         BC   
1597        WNPF               NaN  6/10/2018          USA        NaN   
1605         APF               WPC  17/7/2010          USA         MT   

       MeetTown                           MeetName  
1086  Vancouver        BC Provincial Championships  
1597        NaN  World Tournament Of Championships  
1605        NaN                Big Sky State Games  

[3 rows x 41 columns]

We can use the weightclasses to fill in the first 2 missing bodyweight values. We assume they hold the minimum weight in the particular weightclass

Weight classes can be specified as a maximum or as a minimum. Maximums are specified by just the number, for example `90` means "up to (and including) 90kg." minimums are specified by a `+` to the right of the number, for example `90+` means "above (and excluding) 90kg.

In [238]:
df['BodyweightKg'].fillna(df.WeightClassKg, inplace=True)

In [241]:
df['BodyweightKg'] = df['BodyweightKg'].apply(lambda f: float(f[:-1]) if type(f) == str else f)

In [243]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2793 entries, 0 to 2999
Data columns (total 41 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Name              2793 non-null   object 
 1   Sex               2793 non-null   object 
 2   Event             2793 non-null   object 
 3   Equipment         2793 non-null   object 
 4   Age               2793 non-null   float64
 5   AgeClass          2228 non-null   object 
 6   BirthYearClass    2114 non-null   object 
 7   Division          2793 non-null   object 
 8   BodyweightKg      2792 non-null   float64
 9   WeightClassKg     2780 non-null   object 
 10  Squat1Kg          1074 non-null   float64
 11  Squat2Kg          1066 non-null   float64
 12  Squat3Kg          1044 non-null   float64
 13  Squat4Kg          4 non-null      float64
 14  Best3SquatKg      1850 non-null   float64
 15  Bench1Kg          1696 non-null   float64
 16  Bench2Kg          1687 non-null   float64


The final missing bodyweight value belongs to a Male in the 50-54 AgeClass. We can use the median value of that AgeClass to fill the remaining missing value.

In [246]:
df.groupby(['AgeClass']).median()

Age  BodyweightKg  Squat1Kg  Squat2Kg  Squat3Kg  Squat4Kg  \
AgeClass                                                                
13-15     14.50        58.700     98.75     96.50     83.75      65.0   
16-17     16.50        70.220    115.00    122.50     96.50       NaN   
18-19     18.50        73.900    137.50    147.50    112.50       NaN   
20-23     21.50        81.200    170.00    170.00    118.75     165.0   
24-34     28.00        83.000    147.50    157.50    127.50     241.0   
35-39     37.00        86.910    140.00    117.50    117.50       NaN   
40-44     42.00        91.900    165.00    158.76    117.50       NaN   
45-49     47.00        89.050    123.75    151.25    132.50       NaN   
5-Dec     11.00        39.900     30.00     45.00    -38.75       NaN   
50-54     52.00        82.450    150.00    120.00     82.50       NaN   
55-59     57.00        89.225    110.00    107.50    132.50       NaN   
60-64     61.75        81.890    120.00    135.00     80.00       NaN   
65-69     66.00        74.150     71.25     53.75    -52.50       NaN   
70-74     71.75        85.100   -100.00    105.00    107.50       NaN   
75-79     77.50        80.360     85.00     92.50     96.25       NaN   
80-999    81.50        75.000       NaN       NaN       NaN       NaN   

          Best3SquatKg  Bench1Kg  Bench2Kg  Bench3Kg  ...  Deadlift1Kg  \
AgeClass                                              ...                
13-15          104.330    52.500     57.50    37.500  ...      108.750   
16-17          126.005    70.000     62.50   -35.875  ...      128.750   
18-19          142.500    87.500     90.00   -60.000  ...      170.000   
20-23          186.250   105.000    107.50    56.250  ...      185.000   
24-34          182.500   120.000    115.00   -55.000  ...      190.000   
35-39          193.750   119.000     90.00   -52.500  ...      177.500   
40-44          190.000   132.500    117.50   -67.500  ...      200.000   
45-49          182.500   130.000    105.00   -50.000  ...      150.000   
5-Dec           25.000    23.750     20.00   -22.500  ...       42.215   
50-54          183.750   105.000     96.25   -57.500  ...      185.000   
55-59          150.000    99.895    102.75    60.000  ...      152.500   
60-64          145.000    95.000     83.75   -47.500  ...      162.500   
65-69          130.000    38.750     60.00     1.250  ...      100.000   
70-74          105.000    80.000     90.00    90.000  ...      140.000   
75-79           96.250    72.500     77.50    -3.750  ...      122.500   
80-999         153.750    97.520    102.06   104.330  ...          NaN   

          Deadlift2Kg  Deadlift3Kg  Deadlift4Kg  Best3DeadliftKg  TotalKg  \
AgeClass                                                                    
13-15          118.75       101.05      -105.00           115.00   220.00   
16-17          138.75       130.00      -226.25           137.50   300.00   
18-19          173.75       125.00      -125.00           154.22   331.12   
20-23          190.00       130.00       327.50           212.50   422.50   
24-34          190.00       126.25       125.00           210.00   320.00   
35-39          156.25       142.50       137.50           210.00   267.50   
40-44          203.75       120.00          NaN           218.75   240.00   
45-49          160.00       157.50       188.00           192.64   225.00   
5-Dec           63.50       -40.00       -86.18            35.00    35.00   
50-54          200.00       133.27          NaN           205.00   231.33   
55-59          165.00       155.00          NaN           180.00   326.50   
60-64          180.00       130.00          NaN           185.00   182.50   
65-69          107.50       105.00        85.00           120.00   185.00   
70-74          157.00       162.00          NaN           160.00   288.75   
75-79          130.00        -3.75          NaN           130.00   110.00   
80-999            NaN          NaN          NaN        

In [247]:
df.BodyweightKg.fillna(82.45, inplace=True)